### PR TITLE
state,apiserver, cmd/status: Add MeterStatus to the model

### PR DESCRIPTION
**Pull request by mattyw**
## Description of change

With SLAs operating at the conceptual level of models each model now has the concept of a meter status, this is the same concept as the existing unit meter status.

## QA steps

```bash
juju bootstrap
juju status # No meter status should be reported
# Do something to make the model meter status go red
juju status
# Output should look something like http://paste.ubuntu.com/24196065/
```

## Documentation changes

How users react to meter status on models needs to be documented. This feature is just one of the features under the umbrella of SLAs
## Bug reference

n/a

### CODE REVIEW

File: state/state.go [hunk 1]
```diff
@@ -2172,6 +2172,22 @@ func (st *State) SetSLA(level string, credentials []byte) error {
...
 
+func (st *State) SetModelMeterStatus(status, info string) error {
+	model, err := st.Model()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return model.SetMeterStatus(status, info)
+}
+
+func (st *State) ModelMeterStatus() (MeterStatus, error) {
```
Comment (perrito666) [inline state/state.go]: This method and the previous one are exported but not commented

**Comment by mattyw** [comment]
!!test!!

**Comment by mattyw** [comment]
!!test!!

**Comment by nskaggs** [comment]
!!build!!

File: state/meterstatus.go [hunk 2]
```diff
@@ -76,13 +76,21 @@ type meterStatusDoc struct {
...
-// SetMeterStatus sets the meter status for the unit.
-func (u *Unit) SetMeterStatus(codeStr, info string) error {
+func isValidMeterStatusCode(codeStr string) (MeterStatusCode, error) {
 	code := MeterStatusFromString(codeStr)
 	switch code {
 	case MeterGreen, MeterAmber, MeterRed:
+		return code, nil
 	default:
-		return errors.Errorf("invalid meter status %q", code)
+		return MeterNotAvailable, errors.Errorf("invalid meter status %q", codeStr)
```
Comment (cmars) [inline state/meterstatus.go]: `errors.NotValidf("meter status %q", codeStr)` ?

File: state/upgrades_test.go [hunk 3]
```diff
@@ -100,7 +100,7 @@ func (s *upgradesSuite) TestStripLocalUserDomainModels(c *gc.C) {
 	delete(initialModel, "txn-revno")
 	initialModel["owner"] = "test-admin"
 
-	// TODO (mattyw, cmars) We need to confirm that setting sla in this test
+	// TODO (mattyw, cmars) We need to confirm that setting sla and meter-status in this test
```
Comment (cmars) [inline state/upgrades_test.go]: What is the scenario this test attempts to recreate w.r.t SLAs? Upgrading models from a pre-SLA version of Juju?

**Comment by cmars** [comment]
Also need to consider the case where state has an empty model meter status doc, with an empty string for the status code, if that is the possible after migrating a pre-SLA model. Does that case give an unexpected error when the client apiserver tries to query model meter status?

File: state/upgrades_test.go (hunk 3 above)
Comment (mattyw) [inline state/upgrades_test.go]: @cmars no upgrade is required for SLAs or meter status. This test is actually checking completely seperate behaviour (local domains like "@local" being stripped from user names. However it includes checking against a bson representation of the model, which now includes sla and meter status

**Comment by mattyw** [comment]
@cmars in the case where meter status is queried but non is set we return "NOT AVAILABLE" which is what we do for the unit meter status as well. This isn't bubbled up in juju status as it filters it out.

@cmars regarding combining model & unit meter status. This was being done server side as I understand it

**Comment by mattyw** [comment]
!!test!!

**Comment by mattyw** [comment]
!!build!!

File: apiserver/common/modeldestroy.go [hunk 4]
```diff
@@ -11,13 +11,38 @@ import (
...
+		return nil
+	}
+	return modelBackendShim{st}
+}
+
+type modelBackendShim struct {
+	ModelManagerBackend
+}
+
+func (m modelBackendShim) ForModel(tag names.ModelTag) (metricsender.ModelBackend, error) {
```
Comment (alesstimec) [inline apiserver/common/modeldestroy.go]: godoc

File: apiserver/metricsmanager/metricsmanager.go [hunk 5]
```diff
@@ -176,10 +181,29 @@ func (api *MetricsManagerAPI) SendMetrics(args params.Entities) (params.ErrorRes
...
+		return nil
+	}
+	return modelBackendShim{st}
+}
+
+type modelBackendShim struct {
+	*state.State
+}
+
+func (m modelBackendShim) ForModel(tag names.ModelTag) (metricsender.ModelBackend, error) {
```
Comment (alesstimec) [inline apiserver/metricsmanager/metricsmanager.go]: godoc

File: state/upgrades_test.go (hunk 3 above)
Comment (cmars) [inline state/upgrades_test.go]: @mattyw Thanks for the explanation. I think we can remove this TODO then as this test is unrelated and we're just testing for consistent model migration behavior.

**Comment by mattyw** [comment]
!!build!!

**Comment by cmars** [comment]
@mattyw this one currently has compile errors:
```
# github.com/juju/juju/apiserver/metricsmanager
apiserver/metricsmanager/metricsmanager.go:143: api.state.SLACredential undefined (type metricsender.ModelBackend has no field or method SLACredential)
apiserver/metricsmanager/metricsmanager.go:150: api.state.AllMachines undefined (type metricsender.ModelBackend has no field or method AllMachines)
apiserver/metricsmanager/metricsmanager.go:166: api.state.AddModelMetrics undefined (type metricsender.ModelBackend has no field or method AddModelMetrics)
```

**Comment by cmars** [comment]
!!build!!

**Comment by mattyw** [comment]
!!test!!

**Comment by mattyw** [comment]
!!build!!

**Comment by nskaggs** [comment]
!!build!!

**Comment by nskaggs** [comment]
add to whitelist

**Comment by mattyw** [comment]
!!test!!

**Comment by mattyw** [comment]
$$merge$$

**Comment by cmars** [comment]
@mattyw Test failure is legit:
```
client_test.go:417:
    c.Assert(err, gc.ErrorMatches, test.err)
... error string = "meter status \"NOT AVAILABLE\" not valid"
... regex string = "invalid meter status \"NOT AVAILABLE\""
```

**Comment by mattyw** [comment]
$$merge$$

**Comment by mattyw** [comment]
$$merge$$
